tss-esapi/src/structures/attestation.rs

```diff
@@ -0,0 +1,292 @@
+    type Error = Error;
+
+    fn try_from(tss: TPMI_ST_ATTEST) -> Result<Self> {
+        match tss {
+            TPM2_ST_ATTEST_CERTIFY => Ok(AttestationType::Certify),
+            TPM2_ST_ATTEST_QUOTE => Ok(AttestationType::Quote),
+            TPM2_ST_ATTEST_TIME => Ok(AttestationType::Time),
+            TPM2_ST_ATTEST_CREATION => Ok(AttestationType::Creation),
+            TPM2_ST_ATTEST_NV => Ok(AttestationType::Nv),
+            TPM2_ST_ATTEST_NV_DIGEST => Ok(AttestationType::NvDigest),
+            TPM2_ST_ATTEST_SESSION_AUDIT => Ok(AttestationType::SessionAudit),
+            TPM2_ST_ATTEST_COMMAND_AUDIT => Ok(AttestationType::CommandAudit),
+            _ => Err(Error::local_error(WrapperErrorKind::InvalidParam)),
+        }
+    }
+}
+
+/// Attestation data buffer.
+///
+/// # Details
+/// Corresponds to `TPM2B_ATTEST`. The contents of
+/// the buffer can be unmarshalled into an [Attest]
+/// structure.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct AttestBuffer(Zeroizing<Vec<u8>>);
+
+impl Default for AttestBuffer {
+    fn default() -> Self {
+        AttestBuffer(Vec::new().into())
+    }
+}
+
+impl AttestBuffer {
+    pub const MAX_SIZE: usize = 2304;
+
+    pub fn value(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+impl Deref for AttestBuffer {
+    type Target = Vec<u8>;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl TryFrom<Vec<u8>> for AttestBuffer {
+    type Error = Error;
+
+    fn try_from(bytes: Vec<u8>) -> Result<Self> {
+        if bytes.len() > Self::MAX_SIZE {
+            error!("Error: Invalid Vec<u8> size(> {})", Self::MAX_SIZE);
+            return Err(Error::local_error(WrapperErrorKind::WrongParamSize));
+        }
+        Ok(AttestBuffer(bytes.into()))
+    }
+}
+
+impl TryFrom<&[u8]> for AttestBuffer {
+    type Error = Error;
+
+    fn try_from(bytes: &[u8]) -> Result<Self> {
+        if bytes.len() > Self::MAX_SIZE {
+            error!("Error: Invalid &[u8] size(> {})", Self::MAX_SIZE);
+            return Err(Error::local_error(WrapperErrorKind::WrongParamSize));
+        }
+        Ok(AttestBuffer(bytes.to_vec().into()))
+    }
+}
+
+impl TryFrom<TPM2B_ATTEST> for AttestBuffer {
+    type Error = Error;
+
+    fn try_from(tss: TPM2B_ATTEST) -> Result<Self> {
+        let size = tss.size as usize;
+        if size > Self::MAX_SIZE {
+            error!("Error: Invalid buffer size(> {})", Self::MAX_SIZE);
+            return Err(Error::local_error(WrapperErrorKind::WrongParamSize));
+        }
+        Ok(AttestBuffer(tss.attestationData[..size].to_vec().into()))
+    }
+}
+
+impl TryFrom<AttestBuffer> for Attest {
+    type Error = Error;
+
+    fn try_from(buf: AttestBuffer) -> Result<Self> {
+        let buffer = buf.0.to_vec();
+        let mut attest = TPMS_ATTEST::default();
+        let buf_point: *const u8 = &buffer[0];
+        let mut offset = 0;
+        let ret = Error::from_tss_rc(unsafe {
+            Tss2_MU_TPMS_ATTEST_Unmarshal(
```

Comment:
Hah, see that it was useful of me to add the `MU` sub-library as my first PR ever to this crate? 😄 
I seem to remember someone wondering why it's useful... https://github.com/parallaxsecond/rust-tss-esapi/pull/74#pullrequestreview-420970698 😛

Reply:
Also, this abstraction is great.... If only I'd seen it one day earlier, it could've saved me some time! (https://github.com/keylime/rust-keylime/pull/232/commits/c1cfdb139fdb4f9ff34ff21fd6305c47e626d963#diff-bd3201dff7815da0652c4c02d8d1e9f27f63de6cf5c1d332c2f21ff995ce933cR117)